envoy: Manage life-cycles of singleton maps properly

jrajahalme opened the pull request:
Both policy and host maps are managed as singletons, which are
initialized when first required, and destructed when the last listener
referring to them via Cilium filter instances is removed.

Internally, the maps "post" configuration changes to Envoy worker
threads via thread dispatch queues. These changes are packaged as C++
lambda closures. The problem with this is that while the "posts" are
queued, it is possible for all the listeners to be removed, and thus
the associated singleton maps being freed. If the posted closure
refers to the (now stale) map, bad things can happen.

Fix this by capturing either a weak or shared pointer to the lambda
closure.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

rlenglet commented on envoy/cilium_network_policy.h:
```diff
@@ -25,7 +25,9 @@ class NetworkPolicyMap : public Singleton::Instance,
 		   ThreadLocal::SlotAllocator& tls);
-  ~NetworkPolicyMap() {}
+  ~NetworkPolicyMap() {
+      ENVOY_LOG(info, "Cilium L7 NetworkPolicyMap: NetworkPolicyMap is deleted NOW!");
```
This can be kept at debug level.

jrajahalme committed:
envoy: Manage life-cycles of singleton maps properly.

Both policy and host maps are managed as singletons, which are
initialized when first required, and destructed when the last listener
referring to them via Cilium filter instances is removed.

Internally, the maps "post" configuration changes to Envoy worker
threads via thread dispatch queues. These changes are packaged as C++
lambda closures. The problem with this is that while the "posts" are
queued, it is possible for all the listeners to be removed, and thus
the associated singleton maps being freed. If the posted closure
refers to the (now stale) map, bad things can happen.

Fix this by capturing either a weak or shared pointer to the lambda
closure.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
